### PR TITLE
Replace usages of Marshal.AllocHGlobal with malloc in libraries

### PR DESCRIPTION
**Pull request by reflectronic**
Contributes to https://github.com/dotnet/runtime/issues/54297

Probably best to review commit-by-commit, I tried not to backtrack too much.

Usages in tests will be addressed in a separate PR.

### CODE REVIEW

File: src/libraries/Common/src/System/NativeMemoryHelper.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace System
+{
+    internal static unsafe class NativeMemoryHelper
```
Comment (jkotas) [inline src/libraries/Common/src/System/NativeMemoryHelper.cs:11]: This should be called `NativeMemory` and only included for pre-.NET 6 targets. We do not want to have the wrapper on .NET 6.

Comment (reflectronic) [reply src/libraries/Common/src/System/NativeMemoryHelper.cs:11]: That can work for `Alloc` and `Free`, but it also has some replacements for `Marshal.StringToHGlobalUni` that would have to stay on a type like this. Is that okay?

File: src/libraries/System.Speech/src/Internal/NativeMemorySafeHandle.cs [hunk 2]
```diff
@@ -71,31 +71,26 @@ public override bool IsInvalid
...
         /// </summary>
-        protected override bool ReleaseHandle()
+        protected override unsafe bool ReleaseHandle()
         {
             if (handle != IntPtr.Zero)
             {
-                // Reset the extra information given to the GC
-                if (_bufferSize > 0)
-                {
-                    GC.RemoveMemoryPressure(_bufferSize);
```
Comment (jkotas) [inline src/libraries/System.Speech/src/Internal/NativeMemorySafeHandle.cs:81]: Why did you remove the GC memory pressure call here?

Comment (reflectronic) [reply src/libraries/System.Speech/src/Internal/NativeMemorySafeHandle.cs:81]: Is this a correct application of AddMemoryPressure?

Comment (reflectronic) [reply src/libraries/System.Speech/src/Internal/NativeMemorySafeHandle.cs:81]: Oh, yes, this should not have been removed

File: src/libraries/Common/src/System/NativeMemoryHelper.cs (hunk 1 above)
Comment (jkotas) [reply src/libraries/Common/src/System/NativeMemoryHelper.cs:13]: I would avoid touching StringToHGlobalUni in this PR. If it is useful to have NativeMemory variant of `StringToHGlobalUni` and similar APIs, we should introduce a proper public API for it first. Duplicating NativeMemory variant of `StringToHGlobalUni` via private helper is not an improvement over using StringToHGlobalUni directly.

File: src/libraries/System.Speech/src/Internal/NativeMemorySafeHandle.cs (hunk 2 above)
Comment (jkotas) [reply src/libraries/System.Speech/src/Internal/NativeMemorySafeHandle.cs:81]: I do not know why `System.Speech` is using it. If you would like to clean it up, it should be separate PR.

File: src/libraries/Common/src/System/Security/Cryptography/ECCng.ImportExport.cs [hunk 3]
```diff
@@ -495,8 +495,8 @@ private static ECCurve.ECCurveType ConvertToCurveTypeEnum(Interop.BCrypt.ECC_CUR
                 }
                 finally
                 {
-                    Marshal.FreeHGlobal(descPtr);
-                    Marshal.FreeHGlobal(buffPtr);
+                    NativeMemory.Free((void*)(nint)descPtr);
```
Comment (jkotas) [inline src/libraries/Common/src/System/Security/Cryptography/ECCng.ImportExport.cs:498]: It would be nice to change the type of the variables to `void*` or other appropriate pointer type to make these back and forth casts unnecessary. These casts do not look good.

File: src/libraries/System.Data.Odbc/src/Common/System/Data/ProviderBase/DbBuffer.cs [hunk 4]
```diff
@@ -373,9 +373,9 @@ protected override bool ReleaseHandle()
             // NOTE: The SafeHandle class guarantees this will be called exactly once.
             IntPtr ptr = base.handle;
             base.handle = IntPtr.Zero;
-            if (IntPtr.Zero != ptr)
+            if ((nint)ptr != 0)
```
Comment (jkotas) [inline src/libraries/System.Data.Odbc/src/Common/System/Data/ProviderBase/DbBuffer.cs:376]: These null checks are unnecesary. The Free method contains them already.

File: src/libraries/Common/src/System/NativeMemoryHelper.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/Common/src/System/NativeMemoryHelper.cs:13]: This might be worth profiling more and its probably worth opening a proposal.

On my box, an API which does simply the following, over a `260` character string:
```csharp
var ptr = Marshal.StringToHGlobalUni(s);
Marshal.FreeHGlobal(ptr);
```

Compared to an equivalent API using `NativeMarshal.Alloc/Free` under the covers:
```
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19043.1052 (21H1/May2021Update)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK=6.0.100-preview.5.21302.13
  [Host]     : .NET 6.0.0 (6.0.21.30105), X64 RyuJIT
  DefaultJob : .NET 6.0.0 (6.0.21.30105), X64 RyuJIT

|  Method |     Mean |    Error |   StdDev |
|-------- |---------:|---------:|---------:|
|   Alloc | 41.92 ns | 0.051 ns | 0.045 ns |
| HGlobal | 47.37 ns | 0.071 ns | 0.066 ns |
```

Of course, in comparison to the GC transition and any work the call itself does, this is probably not that meaningful but ~10% overhead can add up.

Comment (tannergooding) [reply src/libraries/Common/src/System/NativeMemoryHelper.cs:13]: (To clarify, I don't think anything needs to be done in this PR, just calling out there is a measurable difference here and so it might be worth opening a proposal because of that. In more extreme cases, many `Alloc` calls can be up to 20x faster than many `LocalAlloc` calls.

File: src/libraries/Common/src/Interop/BSD/System.Native/Interop.Sysctl.cs [hunk 5]
```diff
@@ -54,27 +54,27 @@ private static unsafe void Sysctl(int* name, int name_len, ref byte* value, ref
                 // Do not use ReAllocHGlobal() here: we don't care about
```
Comment (jkotas) [inline src/libraries/Common/src/Interop/BSD/System.Native/Interop.Sysctl.cs:54]: Nit: ReAllocHGlobal

**Comment by reflectronic** [comment]
I think I will split some of the more invasive changes to certain libraries into separate and more comprehensive PRs. Much of this interop code is really old (and really bad), and deserves some serious cleanup in general. Using `NativeMemory` is only a small part of the picture and would only contribute to code uglification in the meantime.

**Comment by reflectronic** [comment]
I am sorry for the slow turnaround time. I have been on vacation this past week and I will finish this up in the next couple of days.

**Comment by jeffhandley** [comment]
@tannergooding This PR is assigned to you for follow-up/decision before the RC1 snap.

**Comment by danmoseley** [comment]
@reflectronic you might prefer to push commits in groups, as each push restarts the build and test.

**Comment by reflectronic** [comment]
Yes; I am sorry, just got a little lazy.
